src/s2e/Plugins/ExecutionTracers/TestCaseGenerator.h

```diff
@@ -21,23 +22,148 @@ namespace s2e {
 namespace plugins {
 namespace testcases {
 
+///
+/// \brief Stores information about file chunks encoded in symbolic variable names
+///
 struct TestCaseFile {
-    /* Maps a chunk id to the size */
+    // Maps a chunk id to the size
     std::map<unsigned, unsigned> chunks;
-    /* Maps a chunk id to the data */
+
+    // Maps a chunk id to the data
     std::map<unsigned, const uint8_t *> chunksData;
+
     unsigned totalParts;
 };
 
-typedef std::map<std::string, TestCaseFile> TestCaseFiles;
+///
+/// \brief Maps a test case file name to a test case file descriptor
+///
+typedef std::unordered_map<std::string, TestCaseFile> TestCaseFiles;
 
+///
+/// \brief Represents any byte-sized data
+///
 typedef std::vector<uint8_t> Data;
+
+///
+/// \brief Maps a test case file name to its concrete data
+///
 typedef std::unordered_map<std::string, Data> TestCaseData;
 
-enum TestCaseType : unsigned { TC_NONE = 0, TC_LOG = 1, TC_TRACE = 2, TC_FILE = 4 };
+///
+/// \brief Specifies the type of the test case to generate
+///
+enum TestCaseType : unsigned {
+    TC_NONE = 0,
 
-/** Handler required for KLEE interpreter */
-class TestCaseGenerator : public Plugin {
+    // The test case will be written to the S2E's debug log
+    TC_LOG = 1,
+
+    // The test case is to be written to the execution trace file
+    TC_TRACE = 2,
+
+    // The test case is a concrete file written to to the output
+    // folder on the host. Symbolic variable names must have a proper
+    // encoding for such test cases to be generated.
+    TC_FILE = 4
+};
+
+///
+/// \brief Maps a sanitized file name to concrete data.
+///
+/// This map will be stored in each execution state and copied
+/// on state fork, because each state may have different sets
+/// of concrete input files.
+///
+/// We use an immutable map in order to avoid needless duplication,
+/// as in general every state will have the exact same copy of the
+/// concrete data.
+///
+typedef klee::ImmutableMap<std::string, Data> ConcreteFileTemplates;
```

Comment:
Shouldn’t the include for this data struct be here and not in the cpp file?

Reply:
Ah wait, I guess it still successfully compiles because you include the Klee header before this one. Let’s change it anyway :)

Reply:
fixed